frontends/p4/simplifyDefUse.cpp

```diff
@@ -1380,20 +1380,22 @@ class FindUninitialized : public Inspector {
         return false;
     }
 
-    bool preorder(const IR::Slice *expression) override {
+    bool preorder(const IR::AbstractSlice *expression) override {
         LOG3("FU Visiting [" << expression->id << "]: " << expression);
 
         auto *slice_stmt = findContext<IR::AssignmentStatement>();
-        if (slice_stmt != nullptr && lhs) {
+        auto *slice = expression->to<IR::Slice>();
+        if (slice_stmt != nullptr && lhs && slice) {
             // track this slice statement
-            hasUses.watchForOverwrites(expression);
+            hasUses.watchForOverwrites(slice);
             LOG4("Tracking " << dbp(slice_stmt) << " " << slice_stmt
                              << " for potential overwrites");
         }
 
         bool save = lhs;
         lhs = false;  // slices on the LHS also read the data
         visit(expression->e0);
+        visit(expression->e1);  // this might not be a constant (for a PlusSlice)
         LOG3("FU Returned from " << expression);
         auto storage = getReads(expression->e0, true);
         reads(expression, storage);  // true even in LHS
```

Comment:
@ChrisDodd Shouldn't `expression` also read `expression->e1` if `expression` is a `IR::PlusSlice`? Something like this:
```
        bool save = lhs;
        lhs = false;  // slices on the LHS also read the data
        visit(expression->e0);
        auto storage = getReads(expression->e0, true);
        reads(expression, storage);  // true even in LHS

        if (expression->is<IR::PlusSlice>()) {
            visit(expression->e1);  // this might not be a constant (for a PlusSlice)
            storage = getReads(expression->e1);
            reads(expression, storage);  // true even in LHS
        }
        LOG3("FU Returned from " << expression);
```

Reply:
> It is not entirely clear to me exactly how this pass works and what it expects for its invariants, but I based this (ie, just setting reads(expression... from expression->e0 and not ->e1) on what is done in the ArrayIndex code a few lines after this. In particular no compound expression ever calls storage->join to create a union of its children.

I noticed the `ArrayIndex` code too, and it also looks suspicious to me, although I don't know what the consequences are of not handling this case like I would expect it to be handled.